### DISCUSSION TITLE
String literals - how express "any character can go here"?

### DISCUSSION THREAD

**Discussion by The-Futurist** [antlr/antlr4#4071] [Q&A]
I'm trying to define a basic `string_literal` rule that can accept:

"dhkjahdkhksfksdfksdf837598ecfeo8t*^*&^BTuyy"; 

as an example.

I don't know how to express the idea of `'"'` `[any-char]*` `'"'` though.

**Comment by The-Futurist** [comment]
I think I found it, just a dot !

**Reply by KvanTTT** [reply]
Also, in this case use can use the following syntax: `'"' ~'"'* '"'`

**Comment by The-Futurist** [comment]
@KvanTTT  - thanks, that's interesting. I am a bit puzzled though. My grammar is parsing this correctly just now:

```
procedure English (X)

	declare counter binary(15);	
 
        counter = FACE B034 D85F BAD7:H;

	if counter > 0 then
	   call retour;
	else
	   return;
	end;

	text = "khsAKJSHakshAJK"
	SHajkhda:QKDHAKSDJAH";

	loop while (a > "aaa")
	   go to place;
	end;
		
end;
```

As you can see it is able to handle, recognize that multi line "string_literal" assignment to "text" halfway down the code, whose rule is:

```
string_literal: basic_string_literal;

basic_string_literal: (DQUOTE ( . | NEWLINE)*? DQUOTE) ;
```
(I'm anticipating additional sub rules for `string_literal` hence the apparent pointless single rule above).

It correctly recognizes the closing quote! It does not get confused by the text `"aaa"` a little further down, it correctly parses the `loop while` and so on.

![image](https://user-images.githubusercontent.com/12262952/212159915-78bc2e13-cba5-45a7-8be9-da86bb9f6b3f.png)

Is it able to do this because it's a _parser_ rule rather than a _lexer_ only rule? It seems too easy, too powerful, many languages have all kinds of fiddly syntax for embedding quotes inside strings yet the above is recognized perfectly - I wonder if my grammar only _seems_ OK, that there might be some hidden gremlin that will bite me later!
